utils/fs-cache.js

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+const { join } = require("path");
+const { mkdir, unlink } = require("fs").promises;
+const { createHash } = require("crypto");
+const { readJSON, writeJSON, readDir } = require("./fs.js");
+const { MemCache } = require("./mem-cache.js");
+
+/**
+ * An in-memory cache persisted by file-system.
+ * @template ValueType
+ */
+class FsCache {
+  /** @type {string} */
+  #dirname = null;
+  /** @type {MemCache<ValueType>} */
+  #cache = null;
+  /** @type {string[]} */
+  #saveQueue = [];
+  /** @type {string[]} */
+  #deleteQueue = [];
+
+  /**
+   * @param {string} dirname where to store cached items on file-system
+   * @typedef {{ key: string, time: number, value: ValueType }} FsCacheEntry
+   */
+  constructor(dirname, ttl = Infinity, autosaveInterval = 10000) {
+    this.#dirname = dirname;
+    this.#cache = new MemCache(ttl);
+    if (autosaveInterval > 0) {
+      setTimeout(this.dump.bind(this), autosaveInterval);
+    }
+  }
+
+  /**
+   * @param {string} key
+   * @param {ValueType} value
+   */
+  set(key, value) {
+    this.#cache.set(key, value);
+    this.#saveQueue.push(key);
+  }
+
+  /**
+   * @param {string} key
+   */
+  async get(key) {
+    const fromMemory = this.#cache.get(key);
+    if (fromMemory) {
+      return fromMemory;
+    }
+    this.#deleteQueue.push(key);
+
+    try {
+      /** @type {FsCacheEntry} */
+      const { time, value } = await readJSON(this.getFilePath(key));
+      this.#cache._set(key, time, value);
+      return value;
+    } catch {
```

Comment:
you sure you want this silent?

Reply:
I think yes. It means, the key was neither in memory nor in file-system. file-system is more like "it's good if it's there, else we'll wait for a `set` anyway"